### DISCUSSION TITLE
Classes which can only be extended internally

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6793] [Language Ideas]
It would be useful to have a sealability modifier which makes it extendable only within the assembly.

This would complement #4032 when it comes to subtype exhaustiveness (both for the compiler à la switch exhaustiveness, and for developers), without the limitations attached to defining all the subtypes as nested types. Some of these benefits may get subsumed by #113, but since DU's would be `enum class`es instead of normal classes, this proposal still has value.

As for the naming:
- `sealed internal` is already taken by `internal sealed`
- I like `sealed internally` despite technically being a new keyword, it doesn't really add any new cognitive strain so it seem OK.

`sealed internally` and `[InternalsVisibleTo]` being in the same assembly would need special support, maybe in the first-pass implementation it can be a compiler error. Both the attribute and `sealed internally` are rare enough that this should be fine in 99.9% of cases (the 0.1% case unsealing the class probably isn't a blocker anyway).

**Comment by Richiban** [comment]
If the class is abstract you can just give it an `internal` constructor; it has the same effect. If you want the type to be instatiable from outside the assembly but not inheritable, you could work around it for now by making its constructor internal and exposing a factory method to construct it.

**Comment by 333fred** [comment]
Duplicate of https://github.com/dotnet/csharplang/issues/485.
